test/run_test.sh

```diff
@@ -0,0 +1,288 @@
+#!/bin/bash
+TS_TMPDIR=
+TS_QUEUE_RESULT=
+TS_EXIT_CODE=0
+TS_LOG_DEBUG=false
+TS_TEST_DIR=
+TS_DISPLAY_TEST_LOG=true
+TS_DISPLAY_TEST_RESULT=true
+TS_SYNC_TEST=false
+TS_TEST_TO_RUN=
+
+help() {
+    cat <<EOF
+$0 [-d][-q][-Q][-s][-o <test_filename>] <test_dir>
+
+Help:
+-d debug log
+-h display this message
+-q quiet tests log
+-Q quiet tests result
+-s run test in a sync way
+-t specify one test filename to run. This file should be in <test_dir>
+EOF
+    exit 0
+}
+
+# Check if all needed environment variable are set
+check_env() {
+    while getopts ":dhqQst:" option
+    do
+        case $option in
+            d)
+                TS_LOG_DEBUG=true
+            ;;
+            h)
+                help
+            ;;
+            q)
+                TS_DISPLAY_TEST_LOG=false
+            ;;
+            Q)
+                TS_DISPLAY_TEST_RESULT=false
+            ;;
+            s)
+                TS_SYNC_TEST=true
+            ;;
+            t)
+                TS_TEST_TO_RUN=$OPTARG
+            ;;
+            :)
+                echo "Arg required for $OPTARG"
+                exit 1
+            ;;
+            \?)
+                echo "Invalid option: $OPTARG"
+                echo
+                help
+            ;;
+        esac
+    done
+    
+    # First args after option should be the TS_TEST_DIR
+    shift $((OPTIND-1))
+    TS_TEST_DIR=$1
+    if [[ -z $TS_TEST_DIR ]];then
+        echo "Test directory not provide"
+        help
+    fi
+    if [[ ! -d $TS_TEST_DIR ]];then
+        log_error "Test directory $TS_TEST_DIR not found"
+    fi
+    
+    # Check test files are OK
+    get_test_files > /dev/null
+}
+
+# Log
+log() {
+    echo "INFO: $@"
+}
+
+log_error() {
+    echo "ERROR: $@" >&2
+    exit 1
+}
+
+log_debug() {
+    [[ $TS_LOG_DEBUG == true ]] || return
+    echo "DEBUG: $@"
+}
+
+# This will enable log with prefix string, this method should be used in a subshell
+enable_prefix_log() {
+    local prefix=$1
+    exec > >(awk '$0="['$prefix'][INFO]\t"$0') 2> >(awk '$0="['$prefix'][ERROR]\t"$0' >&2)
+}
+# reset stdout/stderr to &1 &2
+reset_log() {
+    # restor previsously saved fd
+    exec 1>&6 2>&7
+}
+
+# Disable all output (stdout & stderr)
+disable_log() {
+    exec >/dev/null 2>/dev/null
+}
+
+# This will create all file/queue needed by this test framework
+prepare_system() {
+    TS_TMPDIR=$(mktemp -d)
+    TS_QUEUE_RESULT=$TS_TMPDIR/result
+    touch $TS_QUEUE_RESULT
+    log_debug "tmp dir: $TS_QUEUE_RESULT"
+    
+    # save filedescriptor
+    exec 6>&1 # save stdout in 6
```

Comment:
The whole `fd` juggling is a bit confusing, isn't it overkill if we just want to redirect to `/dev/null` or not depending on a program argument?

Reply:
I understand and I agee but it's mandatory to be able to prefix stdout and stderr in a parallel context. if you have a better solution i'm totally open